app/javascript/mastodon/components/status_list.js

```diff
@@ -53,6 +60,39 @@ class StatusList extends ImmutablePureComponent {
 
   componentWillUnmount () {
     this.detachScrollListener();
+    this.detachIntersectionObserver();
+  }
+
+  attachIntersectionObserver () {
+    const onIntersection = (entries) => {
+      const isIntersecting = { };
+
+      entries.forEach(entry => {
+        const statusId = entry.target.getAttribute('data-id');
+
+        isIntersecting[statusId] = entry.isIntersecting;
+      });
+
+      this.setState(state => ({
+        isIntersecting: Object.assign({ }, state.isIntersecting, isIntersecting),
+      }));
+    };
+
+    const options = {
+      root: this.node,
+      rootMargin: '100% 0px',
+    };
```

Comment:
Another option is to debounce rather than re-rendering when every individual element leaves or enters the viewport.